File: backend/app/schemas/analytics.py
"""
Pydantic DTO Schemas for Procurement Analytics & Impact Dashboard (Part 13)
"""

import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class ProcurementImpactResponse(BaseModel):
    measured_time_reduction_percentage: float
    avg_automated_time_ms: float
    avg_manual_baseline_sec: float
    total_validation_cases: int
    dataset_version: str


class OverviewKPIsResponse(BaseModel):
    total_tenders: int
    active_tenders: int
    total_bids: int
    submitted_bids: int
    evaluated_bids: int
    compliance_rate_percentage: Optional[float] = None
    open_reviews_count: int
    high_critical_risk_bids: int
    average_risk_score: Optional[float] = None
    poor_quality_documents_count: int
    average_quality_score: Optional[float] = None
    duplicate_alerts_count: int
    expired_certificates_count: int = 0
    expiring_soon_certificates_count: int = 0
    procurement_impact: Optional[ProcurementImpactResponse] = None


class FailedRequirementSummary(BaseModel):
    requirement_code: str
    title: str
    category: str
    fail_count: int


class CommonFailureReason(BaseModel):
    reason: str
    count: int


class ComplianceAnalyticsResponse(BaseModel):
    distribution: Dict[str, int]
    total_evaluations: int
    overall_compliance_rate: Optional[float] = None
    mandatory_failures_count: int
    top_failed_requirements: List[FailedRequirementSummary]
    common_failure_reasons: List[CommonFailureReason]


class RiskAnalyticsResponse(BaseModel):
    distribution: Dict[str, int]
    total_risk_evaluated_bids: int
    average_risk_score: Optional[float] = None
    overrides_applied_count: int


class VerificationSourceBreakdown(BaseModel):
    verification_type: str
    total: int
    verified: int
    failed: int
    review_required: int
    success_rate: float


class VerificationAnalyticsResponse(BaseModel):
    status_distribution: Dict[str, int]
    total_verifications: int
    source_breakdown: List[VerificationSourceBreakdown]


class QualityDiagnostics(BaseModel):
    blurry_documents: int
    blank_page_documents: int
    low_resolution_documents: int


class DocumentQualityAnalyticsResponse(BaseModel):
    distribution: Dict[str, int]
    total_documents_analyzed: int
    average_quality_score: Optional[float] = None
    diagnostics: QualityDiagnostics


class DuplicateAnalyticsResponse(BaseModel):
    total_duplicate_alerts: int
    match_type_distribution: Dict[str, int]
    status_distribution: Dict[str, int]


class BulkAnalyticsResponse(BaseModel):
    total_jobs: int
    status_distribution: Dict[str, int]
    total_bids_processed: int
    job_success_rate: Optional[float] = None


class ReviewTypeSummary(BaseModel):
    review_type: str
    count: int


class DisqualificationCategorySummary(BaseModel):
    category: str
    count: int


class HumanReviewAndDecisionResponse(BaseModel):
    total_reviews: int
    review_status_distribution: Dict[str, int]
    review_types_breakdown: List[ReviewTypeSummary]
    total_human_decisions: int
    decision_status_distribution: Dict[str, int]
    disqualification_categories: List[DisqualificationCategorySummary]


class TimeSeriesPoint(BaseModel):
    date: str
    submitted_bids: int
    evaluated_bids: int


class TenderSpecificAnalyticsResponse(BaseModel):
    tender_id: uuid.UUID
    tender_number: str
    title: str
    status: str
    estimated_amount: Optional[float] = None
    overview_kpis: OverviewKPIsResponse
    compliance_analytics: ComplianceAnalyticsResponse
    risk_analytics: RiskAnalyticsResponse
    verification_analytics: VerificationAnalyticsResponse
    human_reviews_and_decisions: HumanReviewAndDecisionResponse

File: frontend/src/types/analytics.ts
/**
 * Procurement Analytics & Impact Types (Part 13)
 */

export interface ProcurementImpact {
  measured_time_reduction_percentage: number;
  avg_automated_time_ms: number;
  avg_manual_baseline_sec: number;
  total_validation_cases: number;
  dataset_version: string;
}

export interface OverviewKPIs {
  total_tenders: number;
  active_tenders: number;
  total_bids: number;
  submitted_bids: number;
  evaluated_bids: number;
  compliance_rate_percentage?: number | null;
  open_reviews_count: number;
  high_critical_risk_bids: number;
  average_risk_score?: number | null;
  poor_quality_documents_count: number;
  average_quality_score?: number | null;
  duplicate_alerts_count: number;
  procurement_impact?: ProcurementImpact | null;
}

export interface FailedRequirementSummary {
  requirement_code: string;
  title: string;
  category: string;
  fail_count: number;
}

export interface CommonFailureReason {
  reason: string;
  count: number;
}

export interface ComplianceAnalytics {
  distribution: Record<string, number>;
  total_evaluations: number;
  overall_compliance_rate?: number | null;
  mandatory_failures_count: number;
  top_failed_requirements: FailedRequirementSummary[];
  common_failure_reasons: CommonFailureReason[];
}

export interface RiskAnalytics {
  distribution: Record<string, number>;
  total_risk_evaluated_bids: number;
  average_risk_score?: number | null;
  overrides_applied_count: number;
}

export interface VerificationSourceBreakdown {
  verification_type: string;
  total: number;
  verified: number;
  failed: number;
  review_required: number;
  success_rate: number;
}

export interface VerificationAnalytics {
  status_distribution: Record<string, number>;
  total_verifications: number;
  source_breakdown: VerificationSourceBreakdown[];
}

export interface QualityDiagnostics {
  blurry_documents: number;
  blank_page_documents: number;
  low_resolution_documents: number;
}

export interface DocumentQualityAnalytics {
  distribution: Record<string, number>;
  total_documents_analyzed: number;
  average_quality_score?: number | null;
  diagnostics: QualityDiagnostics;
}

export interface DuplicateAnalytics {
  total_duplicate_alerts: number;
  match_type_distribution: Record<string, number>;
  status_distribution: Record<string, number>;
}

export interface BulkAnalytics {
  total_jobs: number;
  status_distribution: Record<string, number>;
  total_bids_processed: number;
  job_success_rate?: number | null;
}

export interface ReviewTypeSummary {
  review_type: string;
  count: number;
}

export interface DisqualificationCategorySummary {
  category: string;
  count: number;
}

export interface HumanReviewAndDecision {
  total_reviews: number;
  review_status_distribution: Record<string, number>;
  review_types_breakdown: ReviewTypeSummary[];
  total_human_decisions: number;
  decision_status_distribution: Record<string, number>;
  disqualification_categories: DisqualificationCategorySummary[];
}

export interface TimeSeriesPoint {
  date: string;
  submitted_bids: number;
  evaluated_bids: number;
}

export interface TenderSpecificAnalytics {
  tender_id: string;
  tender_number: string;
  title: string;
  status: string;
  estimated_amount?: number | null;
  overview_kpis: OverviewKPIs;
  compliance_analytics: ComplianceAnalytics;
  risk_analytics: RiskAnalytics;
  verification_analytics: VerificationAnalytics;
  human_reviews_and_decisions: HumanReviewAndDecision;
}
